spanner/spanner_leaderboard/leaderboard.go

```diff
@@ -0,0 +1,348 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command spanner_snippets contains runnable snippet code for Cloud Spanner.
+package main
+
+import (
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"math/rand"
+	"os"
+	"regexp"
+	"strconv"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+
+	"google.golang.org/api/iterator"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+)
+
+type command func(ctx context.Context, w io.Writer, client *spanner.Client) error
+type adminCommand func(ctx context.Context, w io.Writer, adminClient *database.DatabaseAdminClient, database string) error
+
+var (
+	commands = map[string]command{
+		"insertplayers": insertPlayers,
+		"insertscores":  insertScores,
+		"query":         query,
+	}
+)
+
+func createDatabase(ctx context.Context, w io.Writer, adminClient *database.DatabaseAdminClient, db string) error {
+	matches := regexp.MustCompile("^(.*)/databases/(.*)$").FindStringSubmatch(db)
+	if matches == nil || len(matches) != 3 {
+		return fmt.Errorf("Invalid database id %s", db)
+	}
+	op, err := adminClient.CreateDatabase(ctx, &adminpb.CreateDatabaseRequest{
+		Parent:          matches[1],
+		CreateStatement: "CREATE DATABASE `" + matches[2] + "`",
+		ExtraStatements: []string{
+			`CREATE TABLE Players(
+			    PlayerId INT64 NOT NULL,
+			    PlayerName STRING(2048) NOT NULL
+			) PRIMARY KEY(PlayerId)`,
+			`CREATE TABLE Scores(
+			    PlayerId INT64 NOT NULL,
+			    Score INT64 NOT NULL,
+			    Timestamp TIMESTAMP NOT NULL
+			    OPTIONS(allow_commit_timestamp=true)
+			) PRIMARY KEY(PlayerId, Timestamp),
+			INTERLEAVE IN PARENT Players ON DELETE NO ACTION`,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if _, err := op.Wait(ctx); err != nil {
+		return err
+	}
+	fmt.Fprintf(w, "Created database [%s]\n", db)
+	return nil
+}
+
+func insertPlayers(ctx context.Context, w io.Writer, client *spanner.Client) error {
+	// Get number of Players
+	stmt := spanner.Statement{
+		SQL: `SELECT Count(PlayerId) as PlayerCount FROM Players`,
+	}
+	iter := client.Single().Query(ctx, stmt)
+	defer iter.Stop()
+	row, err := iter.Next()
+	if err == iterator.Done {
+		return nil
```

Comment:
Add a comment. Presumably this means that there's already players?

Maybe that's an error?

return errors.New("players already inserted")